Source/WebCore/rendering/svg/RenderSVGResourceClipper.cpp

```diff
@@ -0,0 +1,309 @@
+
+    if (!m_registered) {
+        m_registered = true;
+        registerResource();
+    }
+}
+
+void RenderSVGResourceContainer::idChanged()
+{
+    // Invalidate all our current clients.
+    removeAllClientsFromCache();
+
+    // Remove old id, that is guaranteed to be present in cache.
+    treeScopeForSVGReferences().removeSVGResource(m_id);
+    m_id = element().getIdAttribute();
+
+    registerResource();
+}
+
+void RenderSVGResourceContainer::registerResource()
+{
+    auto& treeScope = this->treeScopeForSVGReferences();
+    if (!treeScope.isIdOfPendingSVGResource(m_id)) {
+        treeScope.addSVGResource(m_id, *this);
+        return;
+    }
+
+    auto elements = copyToVectorOf<Ref<SVGElement>>(treeScope.removePendingSVGResource(m_id));
+
+    treeScope.addSVGResource(m_id, *this);
+
+}
+
+WTF_MAKE_ISO_ALLOCATED_IMPL(RenderSVGResourceClipper);
+
+RenderSVGResourceClipper::RenderSVGResourceClipper(SVGClipPathElement& element, RenderStyle&& style)
+    : RenderSVGResourceContainer(element, WTFMove(style))
+{
+}
+
+RenderSVGResourceClipper::~RenderSVGResourceClipper() = default;
+
+SVGGraphicsElement* RenderSVGResourceClipper::shouldApplyPathClipping() const
+{
+    // If the current clip-path gets clipped itself, we have to fall back to masking.
+    if (style().clipPath())
+        return nullptr;
+
+    SVGGraphicsElement* useGraphicsElement = nullptr;
+
+    // If clip-path only contains one visible shape or path, we can use path-based clipping. Invisible
+    // shapes don't affect the clipping and can be ignored. If clip-path contains more than one
+    // visible shape, the additive clipping may not work, caused by the clipRule. EvenOdd
+    // as well as NonZero can cause self-clipping of the elements.
+    // See also http://www.w3.org/TR/SVG/painting.html#FillRuleProperty
+    for (Node* childNode = clipPathElement().firstChild(); childNode; childNode = childNode->nextSibling()) {
+        auto* renderer = childNode->renderer();
+        if (!renderer)
+            continue;
+        // Only shapes or paths are supported for direct clipping. We need to fall back to masking for texts.
+        if (is<RenderSVGText>(renderer))
+            return nullptr;
+        if (!is<SVGGraphicsElement>(*childNode))
+            continue;
+        auto& style = renderer->style();
+        if (style.display() == DisplayType::None || style.visibility() != Visibility::Visible)
+            continue;
+        // Current shape in clip-path gets clipped too. Fall back to masking.
+        if (style.clipPath())
+            return nullptr;
+        // Fallback to masking, if there is more than one clipping path.
+        if (useGraphicsElement)
+            return nullptr;
+        useGraphicsElement = downcast<SVGGraphicsElement>(childNode);
+    }
+
+    return useGraphicsElement;
+}
+
+enum class ClippingMode {
+    NoClipping,
+    PathClipping,
+    MaskClipping
+};
+
+static ClippingMode& currentClippingMode()
```

Comment:
Hmm, the idea was to avoid switching from PathClipping when already performing MaskClipping...  We should find a nicer way then a shared static variable.

Reply:
Any idea on that one?

Reply:
Let's postpone a cleaner solution - this is at least the fastest, but it would be an issue if we ever do multi-threaded rendering.

Reply:
Sure, I did not find a time to look into it but am not an expert on multi-threading.